### PR TITLE
SYS1 Initializing RPI System

### PR DESCRIPTION
**Pull request by yozohu**
Creating some groundwork for the RPI system, specifically sensor multithreading and inter communication
* Added sensor module, with AbstractSensor class and skeletons for RPI Cam Sensor, Ultrasonic sensor
* Added logging (to console and file) and command line control over log level

### CODE REVIEW

File: bellboy/sensors.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+from bellboy.UltrasonicRanging import pulseIn
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractSensor():
+
+
+class AbstractSensor(ABC):
```
Comment (RyanFleck) [inline bellboy/sensors.py:17]: Codacy found an issue: [expected an indented block (F999)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    ch.setLevel(log_level)
+
+    # create formatter and add it to the handlers
+    formatter = logging.Formatter("%(asctime)s :: %(levelname)s :: %(name)s :: %(funcName)s() :: %(message)s")
+    fh.setFormatter(formatter)
+    ch.setFormatter(formatter)
+
+    # add the handlers to the logger
+    logger.addHandler(fh)
+    logger.addHandler(ch)
```
Comment (RyanFleck) [inline bellboy/bellboy.py:42]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    logger.info(
+        "Logging configured to console and {} at {} level".format(
+            os.path.abspath(log_filename),
+            logging.getLevelName(logger.getEffectiveLevel()),
```
Comment (RyanFleck) [inline bellboy/bellboy.py:46]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py (hunk 1 above)
Comment (RyanFleck) [inline bellboy/sensors.py:17]: Codacy found an issue: [expected an indented block (<string>, line 17)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 3 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:46]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 2 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:42]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ch = logging.StreamHandler()
+    ch.setLevel(log_level)
+
+    # create formatter and add it to the handlers
+    formatter = logging.Formatter("%(asctime)s :: %(levelname)s :: %(name)s :: %(funcName)s() :: %(message)s")
+    fh.setFormatter(formatter)
+    ch.setFormatter(formatter)
+
+    # add the handlers to the logger
+    logger.addHandler(fh)
```
Comment (RyanFleck) [inline bellboy/bellboy.py:41]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    logger.addHandler(ch)
+    logger.info(
```
Comment (RyanFleck) [inline bellboy/bellboy.py:43]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

Comment (RyanFleck) [inline bellboy/bellboy.py:43]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/UltrasonicRanging.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+trigPin = 16
+echoPin = 18
+MAX_DISTANCE = 220  # define the maximum measuring distance, unit: cm
+timeOut = (
+    MAX_DISTANCE * 60
+)  # calculate timeout according to the maximum measuring distance
+
+
+def pulseIn(pin, level, timeOut):  # obtain pulse time of a pin under timeOut
```
Comment (RyanFleck) [inline bellboy/UltrasonicRanging.py:20]: Codacy found an issue: [Redefining name 'timeOut' from outer scope (line 14)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 4 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:41]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/UltrasonicRanging.py (hunk 6 above)
Comment (RyanFleck) [inline bellboy/UltrasonicRanging.py:20]: Codacy found an issue: [Redefining name 'timeOut' from outer scope (line 14)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 5 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:43]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

Comment (RyanFleck) [inline bellboy/bellboy.py:43]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py (hunk 1 above)
Comment (RyanFleck) [inline bellboy/sensors.py:17]: Codacy found an issue: [expected an indented block (<string>, line 17)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 3 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:46]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

Comment (RyanFleck) [inline bellboy/bellboy.py:46]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py (hunk 1 above)
Comment (RyanFleck) [inline bellboy/sensors.py:17]: Codacy found an issue: [expected an indented block (F999)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 4 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:41]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 2 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:42]: Codacy found an issue: [Undefined variable 'logger'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 4 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:41]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py (hunk 2 above)
Comment (RyanFleck) [inline bellboy/bellboy.py:42]: Codacy found an issue: [undefined name 'logger' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

**Comment by yozohu** [comment]
temp closing till ready for review

File: bellboy/sensors.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    def join(self):
+        self._sensor_thread.join()
+
+    def begin(self):
+        logger.debug("starting ultrasonic sensor thread")
+        self._sensor_thread.start()
+
+
+class TestSensor(AbstractSensor):
+    def __init__(self):
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Missing docstring in __init__ (D107)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+class UltrasonicSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
+        super().__init__(poll_period_us)
+
+    # setup the sensor
+    def init_sensor(self, trigPin=16, echoPin=18, max_depth_cm=200, pulse_width_us=10):
+        self._trigPin = trigPin
+        self._echoPin = echoPin
+        self._max_depth_cm = max_depth_cm
+        self._pulse_width = pulse_width_us
```
Comment (RyanFleck) [inline bellboy/sensors.py:182]: Codacy found an issue: [Attribute '_pulse_width' defined outside __init__](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        pass
+
+
+class UltrasonicSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
+        super().__init__(poll_period_us)
+
+    # setup the sensor
+    def init_sensor(self, trigPin=16, echoPin=18, max_depth_cm=200, pulse_width_us=10):
+        self._trigPin = trigPin
```
Comment (RyanFleck) [inline bellboy/sensors.py:179]: Codacy found an issue: [Attribute '_trigPin' defined outside __init__](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        self._echoPin = echoPin
+        self._max_depth_cm = max_depth_cm
```
Comment (RyanFleck) [inline bellboy/sensors.py:181]: Codacy found an issue: [Attribute '_max_depth_cm' defined outside __init__](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+class TestSensor(AbstractSensor):
+    def __init__(self):
+        super().__init__(300)
+
+
+class RPICamSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
+        super().__init__(poll_period_us)
+        self.picam: PiCamera = PiCamera()
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [undefined name 'PiCamera' (F821)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Undefined variable 'PiCamera'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+    # string rep of sensors current status
+    def brief_status(self):
+        pass
+
+    # read sensor once.
+    # to be called continuously in sensor's thread
+    def _read_sensor_loop(self):
+        # send high pulse
+        global logger
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Using the global statement](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+    def brief_status(self):
+        pass
+
+    def read_sensor(self):
+        pass
+
+
+class UltrasonicSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Useless super delegation in method '__init__'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/bellboy.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    # create formatter and add it to the handlers
+    formatter = logging.Formatter("%(asctime)s :: %(levelname)s :: %(name)s :: %(funcName)s() :: %(message)s")
+    fh.setFormatter(formatter)
+    ch.setFormatter(formatter)
+
+    # add the handlers to the logger
+    root_logger.addHandler(fh)
+    root_logger.addHandler(ch)
+    root_logger.info(
+        "Logging configured to console and {} at {} level".format(
```
Comment (RyanFleck) [inline bellboy/bellboy.py:117]: Codacy found an issue: [Use % formatting in logging functions and pass the % parameters as arguments](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            time.sleep(self._pulse_width/1000000.0)  # us --> sec
+            GPIO.output(self._trigPin, GPIO.LOW)  # make trigPin output LOW level
+            pingTime = self.pulseIn(self._echoPin, GPIO.HIGH, self._time_out)   # read plus time of echoPin
+            distance = pingTime * 340.0 / 2.0 / 10000.0
+            self._buffer.append(distance)
+            logger.debug(str.format("Ultrasonic depth recorded: {}", self._buffer[len(self._buffer)-1]))
+            dt_us = (time.time() - t0) * 1000000
+            time.sleep((self._poll_period - dt_us)/1000000)
+
+    def pulseIn(self, pin, level, timeOut):  # obtain pulse time of a pin under timeOut
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Unused argument 'timeOut'](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
+# Sensor classes
+# Sensors include Camera, Ultrasonicm Mic....
+import logging
+import time
+from threading import Thread
+
+import picamera
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: ['picamera' imported but unused (F401)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/UltrasonicRanging.py (hunk 6 above)
Comment (RyanFleck) [inline bellboy/UltrasonicRanging.py:20]: Codacy found an issue: [Redefining name 'timeOut' from outer scope (line 14)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+class AbstractSensor():
+
+    # constructor.
+    # creates private variables for:
+    #   the sensor's data buffer,
+    #   the polling period, and
+    #   the sensor's polling thread
+    # @param poll_period_us time inbetween sensor reads, in microseconds
+    def __init__(self, poll_period_us):
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Missing docstring in __init__ (D107)](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class UltrasonicSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
+        super().__init__(poll_period_us)
+
+    # setup the sensor
+    def init_sensor(self, trigPin=16, echoPin=18, max_depth_cm=200, pulse_width_us=10):
+        self._trigPin = trigPin
+        self._echoPin = echoPin
```
Comment (RyanFleck) [inline bellboy/sensors.py:180]: Codacy found an issue: [Attribute '_echoPin' defined outside __init__](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py (hunk 16 above)
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Unused import picamera](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        self._max_depth_cm = max_depth_cm
+        self._pulse_width = pulse_width_us
+        self._time_out = max_depth_cm*60*0.000001
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Attribute '_time_out' defined outside __init__](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

File: bellboy/sensors.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    def read_sensor(self):
+        pass
+
+
+class UltrasonicSensor(AbstractSensor):
+    def __init__(self, poll_period_us):
+        super().__init__(poll_period_us)
+
+    # setup the sensor
+    def init_sensor(self, trigPin=16, echoPin=18, max_depth_cm=200, pulse_width_us=10):
```
Comment (RyanFleck) [inline bellboy/sensors.py]: Codacy found an issue: [Parameters differ from overridden 'init_sensor' method](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175)

**Comment by RyanFleck** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 171
           

Complexity increasing per file
==============================
- bellboy/bellboy.py  5
- bellboy/LCD/I2CLCD1602.py  2
- bellboy/sensors.py  4
- bellboy/UltrasonicRanging.py  5
- bellboy/LCD/UltrasonicRanging.py  5
- bellboy/LCD/Adafruit_LCD1602.py  7
- bellboy/test/test_system.py  2
- bellboy/LCD/PCF8574.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Bellboy-Capstone/System/pullRequest?prid=6272175&bid=20416699)

**Pull request closed without merging**